### PR TITLE
Obtain entity group lock during commit

### PR DESCRIPTION
**Pull request by cdonati**
This avoids long-running locks obtained during the course of a transaction. It limits concurrency errors to commit operations.

### CODE REVIEW

File: AppDB/appscale/datastore/zkappscale/entity_lock.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+from kazoo.retry import (
+  ForceRetryError,
+  KazooRetry,
+  RetryFailedError
+)
+
+# The ZooKeeper node that contains lock entries for an entity group.
+LOCK_PATH_TEMPLATE = '/appscale/apps/{project}/locks/{namespace}/{group}'
+
+# The number of seconds to wait before
```
Comment (menivaitsi) [inline AppDB/appscale/datastore/zkappscale/entity_lock.py]: Incomplete comment.

File: AppDB/appscale/datastore/zkappscale/entity_lock.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    for path in self.paths:
+      self.client.ensure_path(path)
+
+  def cancel(self):
+    """ Cancel a pending lock acquire. """
+    self.cancelled = True
+    self.wake_event.set()
+
+  def acquire(self):
+    """ Acquire the lock. By defaults blocks and waits forever.
```
Comment (menivaitsi) [inline AppDB/appscale/datastore/zkappscale/entity_lock.py]: defaults -> default

File: AppDB/appscale/datastore/zkappscale/zktransaction.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from ..dbconstants import (MAX_GROUPS_FOR_XG,
                            MAX_TX_DURATION)
 from ..unpackaged import APPSCALE_PYTHON_APPSERVER
+from kazoo.exceptions import NoNodeError
```
Comment (menivaitsi) [inline AppDB/appscale/datastore/zkappscale/zktransaction.py]: Move this one line down.

File: AppDB/appscale/datastore/cassandra_env/cassandra_interface.py [hunk 4]
```diff
@@ -1023,8 +1080,8 @@ def delete_entities_tx(self, app, txid, entity_keys):
     insert = self.session.prepare("""
       INSERT INTO transactions (txid_hash, operation, namespace, path, entity)
       VALUES (?, ?, ?, ?, ?)
-      USING TTL 120
-    """)
+      USING TTL {ttl}
+    """.format(ttl=dbconstants.MAX_TX_DURATION * 2))
```
Comment (cdonati) [inline AppDB/appscale/datastore/cassandra_env/cassandra_interface.py]: It just gives some reasonable padding to ensure the state is present for the duration of the transaction. Eventually, we may need it to exist a bit longer than the transaction for the taskqueue to query the status of a transaction.

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/2498/

**Comment by cdonati** [comment]
Blobstore build: https://ocd.appscale.com:8080/job/Blobstore/1602/

**Pull request merged**